Repeat decorator

ijassiem opened the pull request.

ijassiem committed:
Add decorator-function for repeating methods in Pokemon class

This decorator @repeat provides the functionality of repeating the execution
of a method a random number of times, if the corresponding method is decorated.
The random number is generated ranging from 0-10. This decorator is used with
speak method in the Pokemon class. A unit test is included for ensuring that the
number of times the speak method is repeated does not exceed maximum number of
repeats which is 10.

james-smith-za commented on pokemon/pokemon.py:
```diff
@@ -0,0 +1,94 @@
+
+    def inner(func_object):
+        def wrapper(*args, **kwargs):
+            print("Repeat ", m, " times")
```
Python f-strings are a cool thing to use, you may want to look into them:
https://realpython.com/python-f-strings/
```suggestion
            print(f"Repeat {m} times")
```

james-smith-za commented on pokemon/pokemon.py:
```diff
@@ -0,0 +1,94 @@
+        self.nickname = nickname
+        self.moves = ["jump", "strike", "dash", "block"]
+
+    @repeat(randint(0, repeat_max))
+    def speak(self):
+        """Pokemon say name and repeat random times."""
+        print(self.name)
```
So you've got the general idea here, but what will happen with this is the number will be chosen randomly once, and then forever more you'll be repeating the same number of times.

I'm not sure whether it'll be set for the class as a whole or each instance of the class, I'd need to sit and think about the inner workings, but I think it's the former.

What I had in mind was to repeat the function (i.e. printing the message) a different number of times, each time the `speak()` function is called.

Have a look at the code that I came up with earlier, and see where I use the `randint()` method. Try to see if you can understand why the difference is there:

```
from random import randint

def repeat_randomly(n=5):
  def decorator_repeat_randomly(func):
    def wrapper(*args, **kwargs):
      for _ in range(randint(1,n)):
        value = func(*args, **kwargs)
      return value
    return wrapper
  return decorator_repeat_randomly

@repeat_randomly(6)
def print_a_message(message):
  print(message)
```

ijassiem committed:
Implement changes relating to comments in pull request.

These changes pertains to the comments indicated in the pull request: "Repeat decorator #4".

Added the use of f-strings with print(). Changed the decorator-function to generate a random number from
inside the function, resulting in a new random number being generated each time the speak() method is called.

Contributes to: DGDT-26

ijassiem committed:
Merge branch 'main' into repeat_decorator
